meadow-client/src/MonadApp.purs

```diff
@@ -0,0 +1,394 @@
+import Effect (Effect)
+import Effect.Aff.Class (class MonadAff)
+import Effect.Class (class MonadEffect)
+import FileEvents as FileEvents
+import Gist (Gist, GistId, NewGist)
+import Halogen (HalogenM, liftAff, liftEffect, query', request)
+import Language.Haskell.Interpreter (InterpreterError, InterpreterResult, SourceCode)
+import LocalStorage as LocalStorage
+import Marlowe.Parser (contract)
+import Marlowe.Semantics
+  ( ErrorResult(InvalidInput)
+  , IdInput(IdOracle, InputIdChoice)
+  , MApplicationResult(MCouldNotApply, MSuccessfullyApplied)
+  , OracleDataPoint(..)
+  , State(State)
+  , TransactionOutcomes
+  , applyTransaction
+  , collectNeededInputs
+  , emptyState
+  , peopleFromStateAndContract
+  , reduce
+  , scoutPrimitives
+  )
+import Marlowe.Types (BlockNumber, Choice, Contract(Null), IdChoice(IdChoice), IdOracle, Person, WIdChoice(WIdChoice))
+import Meadow (SPParams_)
+import Meadow as Server
+import Network.RemoteData as RemoteData
+import Servant.PureScript.Ajax (AjaxError)
+import Servant.PureScript.Settings (SPSettings_)
+import StaticData (bufferLocalStorageKey, marloweBufferLocalStorageKey)
+import Text.Parsing.Parser (runParser)
+import Types
+  ( ChildQuery
+  , ChildSlot
+  , EditorSlot(EditorSlot)
+  , FrontendState
+  , InputData
+  , MarloweEditorSlot(MarloweEditorSlot)
+  , MarloweState
+  , OracleEntry
+  , Query
+  , TransactionData
+  , TransactionValidity(..)
+  , WebData
+  , _choiceData
+  , _contract
+  , _currentMarloweState
+  , _input
+  , _inputs
+  , _marloweState
+  , _oldContract
+  , _oracleData
+  , _outcomes
+  , _signatures
+  , _transaction
+  , _validity
+  , cpEditor
+  , cpMarloweEditor
+  )
+import Web.HTML.Event.DragEvent (DragEvent)
+
+class
+  Monad m <= MonadApp m where
+  editorSetValue :: String -> Maybe Int -> m Unit
+  editorGetValue :: m (Maybe String)
+  editorSetAnnotations :: Array Annotation -> m Unit
+  editorGotoLine :: Int -> Maybe Int -> m Unit
+  marloweEditorSetValue :: String -> Maybe Int -> m Unit
+  marloweEditorGetValue :: m (Maybe String)
+  preventDefault :: DragEvent -> m Unit
+  readFileFromDragEvent :: DragEvent -> m String
+  updateContractInState :: String -> m Unit
+  updateState :: m Unit
+  saveInitialState :: m Unit
+  updateMarloweState :: (MarloweState -> MarloweState) -> m Unit
+  resetContract :: m Unit
+  saveBuffer :: String -> m Unit
+  saveMarloweBuffer :: String -> m Unit
+  getOauthStatus :: m (WebData AuthStatus)
+  getGistByGistId :: GistId -> m (WebData Gist)
+  postGist :: NewGist -> m (WebData Gist)
+  patchGistByGistId :: NewGist -> GistId -> m (WebData Gist)
+  postContractHaskell :: SourceCode -> m (WebData (JsonEither InterpreterError (InterpreterResult RunResult)))
+
+newtype HalogenApp m a
+  = HalogenApp (HalogenM FrontendState Query ChildQuery ChildSlot Void m a)
+
+derive instance newtypeHalogenApp :: Newtype (HalogenApp m a) _
+
+instance functorHalogenApp :: Functor (HalogenApp m) where
+  map f m = wrap $ map f $ unwrap m
```

Comment:
Oh look - you've just given me a great idea!

https://github.com/input-output-hk/plutus/pull/1118

Reply:
done